### PR TITLE
Create memory bin override library to allow for more extensive platform customization 

### PR DESCRIPTION
**Pull request by cfernald**
## Description

Adds a memory bin override library that allows a platform to customize logic around memory bin locations in DXE.

- [ ] Impacts functionality?
- [ ] Impacts security?
- [x] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested

Tested on the Q35 platform with a custom memory bin override library

## Integration Instructions

Add the new NULL version of the library to the platform DSC file
```
MemoryBinOverrideLib|MdeModulePkg/Library/MemoryBinOverrideLibNull/MemoryBinOverrideLibNull.inf
```

### CODE REVIEW

**Review by os-d — APPROVED** [review]

File: MdeModulePkg/Core/Dxe/Mem/Page.c
Comment (apop5) [inline MdeModulePkg/Core/Dxe/Mem/Page.c:625]: is the base address necessary here? 

it gets passed into CheckMemoryBinOverride, but then in CoreAllocatepages below, it will be overwritten...

Comment (apop5) [reply MdeModulePkg/Core/Dxe/Mem/Page.c:625]: Oh, Depending on the allocation type returned, it might be requesting a specific address.

**Review by apop5 — APPROVED** [review]

**Comment by makubacki** [comment]
@cfernald have you and @kenlautner discussed how this and https://github.com/microsoft/mu_basecore/pull/167 would cohesively work together. Seems like a design plan should take both changes into account.

**Comment by cfernald** [comment]
> @cfernald have you and @kenlautner discussed how this and #167 would cohesively work together. Seems like a design plan should take both changes into account.

Agreed. This change will likely be over-hauled after merging the PEI phase bucket work. Due to this change being needed to unblock other work however, I will come back and do that work after both of these changes are merged.

**Comment by kenlautner** [comment]
If this change only cares about the DXE memory buckets then with how my PR is currently structured it shouldn't affect it.  The DXE memory buckets don't really care about what the PEI memory looks like.

File: MdeModulePkg/Include/Library/MemoryBinOverrideLib.h
Comment (makubacki) [inline MdeModulePkg/Include/Library/MemoryBinOverrideLib.h]: Library class header needs to be added to MdeModulePkg.dec

File: MdeModulePkg/Library/MemoryBinOverrideLibNull/MemoryBinOverrideLibNull.c
Comment (makubacki) [inline MdeModulePkg/Library/MemoryBinOverrideLibNull/MemoryBinOverrideLibNull.c]: Was `DebugLib` really needed in this file?

**Comment by makubacki** [comment]
> If this change only cares about the DXE memory buckets then with how my PR is currently structured it shouldn't affect it. The DXE memory buckets don't really care about what the PEI memory looks like.

I'm not so concerned about whether the two can coexist but that that they present cohesive interfaces. For example, this PR adds `MemoryBinOverrideLib` and the other PR adds `MemoryBucketLib`. Do we really need to call mix & match bin & bucket terminology?

File: MdeModulePkg/Include/Library/MemoryBinOverrideLib.h
Comment (kuqin12) [inline MdeModulePkg/Include/Library/MemoryBinOverrideLib.h]: Duplicated `OUT`?

**Review by kuqin12 — APPROVED** [review]

File: MdeModulePkg/Include/Library/MemoryBinOverrideLib.h
Comment (makubacki) [inline MdeModulePkg/Include/Library/MemoryBinOverrideLib.h]: I think `GetMemoryBinOverride()` would be more precise to what's happening here than check since you're getting back the override values rather than a bool indicating if an override is needed.

**Review by makubacki — APPROVED** [review]

**Comment by cfernald** [comment]
> > If this change only cares about the DXE memory buckets then with how my PR is currently structured it shouldn't affect it. The DXE memory buckets don't really care about what the PEI memory looks like.
> 
> I'm not so concerned about whether the two can coexist but that that they present cohesive interfaces. For example, this PR adds `MemoryBinOverrideLib` and the other PR adds `MemoryBucketLib`. Do we really need to call mix & match bin & bucket terminology?

I think this is fair. I intend to update the functionality here once we have PEI memoru buckets as those will also be accounted for by the platform. For now I will leave it as it to and will update post check-in of the PEI memory buckets.

**Comment by makubacki** [comment]
> > > If this change only cares about the DXE memory buckets then with how my PR is currently structured it shouldn't affect it. The DXE memory buckets don't really care about what the PEI memory looks like.
> > 
> > 
> > I'm not so concerned about whether the two can coexist but that that they present cohesive interfaces. For example, this PR adds `MemoryBinOverrideLib` and the other PR adds `MemoryBucketLib`. Do we really need to call mix & match bin & bucket terminology?
> 
> I think this is fair. I intend to update the functionality here once we have PEI memoru buckets as those will also be accounted for by the platform. For now I will leave it as it to and will update post check-in of the PEI memory buckets.

Sounds good. We technically should reduce thrash of public interfaces (like decide on bin vs buckets before the interface gets introduced). But I mainly wanted to raise the idea that we should make sure these things go together well. For what it's worth, I marginally prefer bucket simply because "bin" could be confused with "binary" for the unaware. But not a big deal either way.

**Comment by cfernald** [comment]
> > > > If this change only cares about the DXE memory buckets then with how my PR is currently structured it shouldn't affect it. The DXE memory buckets don't really care about what the PEI memory looks like.
> > > 
> > > 
> > > I'm not so concerned about whether the two can coexist but that that they present cohesive interfaces. For example, this PR adds `MemoryBinOverrideLib` and the other PR adds `MemoryBucketLib`. Do we really need to call mix & match bin & bucket terminology?
> > 
> > 
> > I think this is fair. I intend to update the functionality here once we have PEI memoru buckets as those will also be accounted for by the platform. For now I will leave it as it to and will update post check-in of the PEI memory buckets.
> 
> Sounds good. We technically should reduce thrash of public interfaces (like decide on bin vs buckets before the interface gets introduced). But I mainly wanted to raise the idea that we should make sure these things go together well. For what it's worth, I marginally prefer bucket simply because "bin" could be confused with "binary" for the unaware. But not a big deal either way.

I too prefer bucket for this reason, though I chose bin here as page.c refers to these as bins throughout currently.

**Pull request merged**
